thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/classification/ClassificationTaskRunner.java

```diff
@@ -1,17 +1,161 @@
+  }
+
+  private void runTask(long windowStart, long windowEnd) {
+    long mainFunctionId = classificationConfig.getMainFunctionId();
+    addAnomalyFunctionAndAlertConfig(mainFunctionId);
+    AlertFilter alertFilter = alertFilterMap.get(mainFunctionId);
+
+    // Get the anomalies from the main anomaly function
+    List<MergedAnomalyResultDTO> mainAnomalies =
+        mergedAnomalyDAO.findAllOverlapByFunctionId(mainFunctionId, windowStart, windowEnd, false);
+    List<MergedAnomalyResultDTO> filteredMainAnomalies = filterAnomalies(alertFilter, mainAnomalies);
+
+    // Sort merged anomalies by the natural order of the end time
+    Collections.sort(filteredMainAnomalies, new MergeAnomalyEndTimeComparator());
+    // Run classifier for each dimension of the anomalies
+    ListMultimap<DimensionMap, MergedAnomalyResultDTO> updatedMainAnomaliesByDimension =
+        dimensionalShuffleAndUnifyClassification(filteredMainAnomalies);
+
+  }
+
+  private ListMultimap<DimensionMap, MergedAnomalyResultDTO> dimensionalShuffleAndUnifyClassification(
+      List<MergedAnomalyResultDTO> mainAnomalies) {
+
+    ListMultimap<DimensionMap, MergedAnomalyResultDTO> updatedMainAnomaliesByDimension = ArrayListMultimap.create();
+    // Terminate if the main anomaly function has not detected any anomalies in the given window
+    if (CollectionUtils.isEmpty(mainAnomalies)) {
+      return updatedMainAnomaliesByDimension;
+    }
+
+    // Sort anomalies by their dimensions
+    ListMultimap<DimensionMap, MergedAnomalyResultDTO> mainAnomaliesByDimension = ArrayListMultimap.create();
+    for (MergedAnomalyResultDTO mainAnomaly : mainAnomalies) {
+      mainAnomaliesByDimension.put(mainAnomaly.getDimensions(), mainAnomaly);
+    }
+
+    // Set up maps of function id to anomaly function config and alert filter
+    List<Long> functionIds = classificationConfig.getFunctionIdList();
+    for (Long functionId : functionIds) {
+      addAnomalyFunctionAndAlertConfig(functionId);
+    }
+
+    // For each dimension, we get the anomalies from the correlated metric
+    for (DimensionMap dimensionMap : mainAnomaliesByDimension.keySet()) {
+      Map<Long, List<MergedAnomalyResultDTO>> functionIdToAnomalyResult = new HashMap<>();
+
+      // Get the anomalies from other anomaly function that are activated
+      for (Long functionId : functionIds) {
+        AnomalyFunctionDTO anomalyFunctionDTO = anomalyFunctionConfigMap.get(functionId);
+        AlertFilter alertFilter = alertFilterMap.get(functionId);
+        if (anomalyFunctionDTO.getIsActive()) {
+          List<MergedAnomalyResultDTO> anomalies = mergedAnomalyDAO
+              .findAllOverlapByFunctionIdDimensions(functionId, windowStart, windowEnd, dimensionMap.toString(), false);
+          List<MergedAnomalyResultDTO> filteredAnomalies = filterAnomalies(alertFilter, anomalies);
+          Collections.sort(filteredAnomalies, new MergeAnomalyEndTimeComparator());
```

Comment:
can we use just 1 MergeAnomalyEndTimeComparator instance instead of creating a new object in every iteration?